packages/babel-preset-expo/src/server-data-loaders-plugin.ts

```diff
@@ -0,0 +1,108 @@
+/**
+ * Copyright © 2025 650 Industries.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+import type { ConfigAPI, types as t, PluginObj } from '@babel/core';
+
+import { getExpoRouterAbsoluteAppRoot, getIsServer, toPosixPath } from './common';
+
+const debug = require('debug')('expo:babel:server-data-loaders');
+
+const LOADER_EXPORT_NAME = 'loader';
+
+export function serverDataLoadersPlugin(api: ConfigAPI & typeof import('@babel/core')): PluginObj {
+  const { types: t } = api;
+
+  const routerAbsoluteRoot = api.caller(getExpoRouterAbsoluteAppRoot);
+  const isServer = api.caller(getIsServer);
+
+  return {
+    name: 'expo-server-data-loaders',
+    visitor: {
+      ExportNamedDeclaration(path, state) {
+        assertExpoMetadata(state.file.metadata);
+
+        if (isServer) {
+          return;
+        }
+
+        // Early exit if file is not within the `app/` directory
+        if (!isInAppDirectory(state.file.opts.filename ?? '', routerAbsoluteRoot)) {
+          debug('Skipping file outside app directory:', state.file.opts.filename);
+          return;
+        }
+
+        // Early exit if file doesn't contain a `loader` named export
+        if (!state.file.code.includes(LOADER_EXPORT_NAME)) {
+          debug('Skipping file:', state.file.opts.filename);
+          state.file.path.stop();
+          return;
+        }
+
+        debug(
+          isServer ? 'Processing server bundle:' : 'Processing client bundle:',
+          state.file.opts.filename
+        );
+
+        const { declaration } = path.node;
```

Comment:
This should bail if `declaration.exportKind === 'type'`

We also need to handle `specifiers`. Technically, this is mutually exclusive (a node can either have `specifiers` or `declaration`, but Babel doesn't enforce this strictly (just loosely, I think during generation and creation).

We should bail on `specifiers`

I never tested what happens when `declaration` is `path.remove()`'d and both `specifiers` and `declaration` are missing. Do we have a test for this?